### PR TITLE
Features/tryon

### PR DESCRIPTION
**Pull request by zouxinyi0625**
Add tryon as a postprocess for EasyPhoto.
For more details, you can refer to our wiki: https://github.com/aigc-apps/sd-webui-EasyPhoto/wiki/TryOn.
UI:
1. upload a template image (and give hints of the template ROI)
2. choose from cloth gallery (Your cloth will also be added into cloth gallery) / upload a reference image (and give hints of the reference ROI)
3. Click "Start Generation"
image

### CODE REVIEW

File: scripts/easyphoto_ui.py [hunk 1]
```diff
@@ -110,1252 +112,1791 @@ def get_block_name(self):
...
-                                )
+                            with gr.Accordion("Templtate Image", open=True) as tryon_template_chosen:
                                 with gr.Row():
-                                    t2v_input_width = gr.Slider(
-                                        minimum=64, maximum=2048, step=8, label="Video Width", value=512, elem_id=f"width"
-                                    )
-                                    t2v_input_height = gr.Slider(
-                                        minimum=64, maximum=2048, step=8, label="Video Height", value=768, elem_id=f"height"
+                                    template_upload_way = gr.Radio(
+                                        ["Template Gallery", "Template Upload"],
```
[the remark this answers is not in the document]
Comment (wuziheng) [reply scripts/easyphoto_ui.py:1414]: 需要进一步考虑一下布局~ need fix

File: scripts/easyphoto_utils/tryon_utils.py [hunk 2]
```diff
@@ -0,0 +1,940 @@
...
+    # Fit a rotated rectangle around the polygon
+    rect = cv2.minAreaRect(points)
+
+    # Get the rotation angle of the minimum bounding box
+    rotation_angle = rect[2]
+
+    return rotation_angle
+
+
+def find_best_angle_ratio(
```
Comment (wuziheng) [inline scripts/easyphoto_utils/tryon_utils.py:389]: someday ，we will finish a match function for 3dmesh with pitch/yaw/roll

File: scripts/easyphoto_utils/tryon_utils.py [hunk 3]
```diff
@@ -0,0 +1,940 @@
...
+    constraint = {"type": "ineq", "fun": constraint_function}
+
+    # Define optimization options, including learning rate
+    options = {"disp": True, "maxiter": max_iters, "ftol": 1e-6, "eps": 1e-5}
+
+    # Define parameter bounds
+    bounds = [(angle_target - 10, angle_target + 10), (0.1, 3.0)]
+
+    # Perform the optimization to maximize IoU while satisfying the constraint
+    result = minimize(
```
Comment (wuziheng) [inline scripts/easyphoto_utils/tryon_utils.py:493]: 宣传文章可以出一个示意图，以及提供一些debug的内容？

File: scripts/easyphoto_infer.py [hunk 4]
```diff
@@ -300,7 +330,26 @@ def inpaint(
     h = int(input_image.height) if type(input_image) is not list else int(input_image[0].height)
 
     for pair in controlnet_pairs:
-        controlnet_units_list.append(get_controlnet_unit(pair[0], pair[1], pair[2], False if type(input_image) is not list else True))
+        if len(pair) == 4:
+            # set control_mode
+            controlnet_units_list.append(
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_infer.py:335]: done

File: scripts/easyphoto_tryon_infer.py [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+    if num_connected < 2:
+        ep_logger.info(f"{(img_type)} Refine input mask of by mask.")
+        # support the input is a mask, we use box and sam to refine mask
+        _, box_template = mask_to_box(mask[:, :, 0])
+        mask = np.uint8(seg_by_box(np.array(img), box_template, predictor))
+    else:
+        ep_logger.info(f"{(img_type)} Refine input mask of by points.")
+        # support points is given, points are used to refine mask
+        centroids = np.array(centroids)
+        input_label = np.array([1] * centroids.shape[0])
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_tryon_infer.py:708]: 支持两种方式：
1. 通过多个point （SAM 的输入是point）
image

2. 通过显示的画一个区域/Mask （SAM 的输入是这个区域得到的box）（Mask 只有一个联通区域的时候触发）
image

File: scripts/easyphoto_tryon_infer.py [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+    if checkpoint_type == 2 or checkpoint_type == 3:
+        info = "Tryon does not support the SD2 / SDXL checkpoint."
+        ep_logger.error(info)
+        return info, [], template_mask, reference_mask
+
+    cloth_gallery_dir = os.path.join(cloth_id_outpath_samples, "gallery")
+    gallery_lists = glob(os.path.join(cloth_gallery_dir, "*.jpg")) + glob(os.path.join(cloth_gallery_dir, "*.png"))
+    user_ids = [i.split("/")[-1].split(".")[0] for i in gallery_lists]
+    ep_logger.info(f"user_ids: {user_ids}")
+
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_tryon_infer.py:103]: done

File: scripts/easyphoto_tryon_infer.py [hunk 7]
```diff
@@ -0,0 +1,695 @@
...
+    return_msg = ""
+
+    webui_save_path = os.path.join(models_path, f"Lora/{cloth_uuid}.safetensors")
+    if os.path.exists(webui_save_path):
+        return_msg += f"Use exists LoRA of {cloth_uuid}.\n"
+        ep_logger.info(f"LoRA of user id: {cloth_uuid} exists. Start Infer.")
+    else:
+
+        def has_duplicate_uid(input_user_id, existing_user_ids):
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_tryon_infer.py:182]: 对，实际的cloth_id 是 填写的cloth_id+设置的train steps， 一件衣服的不同train_step也会视为不同的uid。

File: scripts/easyphoto_tryon_infer.py [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+        return_res.append(first_paste)
+        torch.cuda.empty_cache()
+
+        ep_logger.info("Finished")
+
+    return "Success\n" + return_msg, return_res, template_mask, reference_mask
+
+
+def easyphoto_tryon_mask_forward(input_image, img_type):
+    global predictor
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_tryon_infer.py]: done

File: scripts/easyphoto_infer.py [hunk 9]
```diff
@@ -340,7 +389,8 @@ def inpaint(
 face_skin = None
 face_recognition = None
 psgan_inference = None
-check_hash = [True, True, True, True, True, True]
+# base, portrait, sdxl, add_text2image, add_ipa_base, add_ipa_sdxl, add_video
+check_hash = [True, True, True, True, True, True, True]
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_infer.py:393]: 对

File: scripts/easyphoto_ui.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_ui.py:1416]: upload UI with larger height.

[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply scripts/easyphoto_ui.py:1416]: 换了个大框，好看一点了。

File: scripts/easyphoto_tryon_infer.py [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+            ]
+            try:
+                subprocess.run(command, check=True)
+            except subprocess.CalledProcessError as e:
+                ep_logger.error(f"Error executing the command: {e}")
+
+        best_weight_path = os.path.join(weights_save_path, f"pytorch_lora_weights.safetensors")
+        if not os.path.exists(best_weight_path):
+            return ("Failed to obtain Lora after training, please check the training process.", [], template_mask, reference_mask)
+
```
Comment (RetrieverJo) [inline scripts/easyphoto_tryon_infer.py:336]: I got an error occurred here because the target folder does not exist. Following suggested code should be added.

```suggestion
        os.makedirs(cloth_gallery_dir, exist_ok=True)
```

**Review by RetrieverJo — CHANGES_REQUESTED** [review]

File: scripts/easyphoto_tryon_infer.py (hunk 10 above)
Comment (wuziheng) [reply scripts/easyphoto_tryon_infer.py:336]: OK，thanks

**Comment by wuziheng** [comment]
> When I ran it for the first time, I used a custom model(fine-tuned on SD1.5) other than the default model(Chilloutmix), but an error occurred. That error was validating hash of base model in the inference time (I guess). The inference step tried to validate with default model and my custom model.
> 
> Unfortunately, now it is impossible to reproduce 😢

maybe run chillout at first time to pass hash check is OK

**Comment by wuziheng** [comment]
tested  pass
1. color problem
2. more model

gallery model + gallery cloth
upload model + gallery cloth
gallery model + upload cloth

**Comment by wuziheng** [comment]
t3
t2
t1

File: scripts/easyphoto_infer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wuziheng) [reply scripts/easyphoto_infer.py:335]: pair=4 means  [control_model_name, xx, control_weight, control_mode], eg ['canny', xx, 1.0, 0], 0 means xx

**Review by wuziheng — APPROVED** [review]
LGTM，after fix windows training bug

File: scripts/easyphoto_tryon_infer.py [hunk 11]
```diff
@@ -0,0 +1,721 @@
...
+                f"--max_train_steps={max_train_steps}",
+                f"--checkpointing_steps={val_and_checkpointing_steps}",
+                f"--learning_rate={learning_rate}",
+                "--lr_scheduler=constant",
+                "--lr_warmup_steps=0",
+                "--train_text_encoder",
+                "--seed=42",
+                f"--rank={rank}",
+                f"--network_alpha={network_alpha}",
+                f"--output_dir={os.path.relpath(weights_save_path, pwd)}",
```
Comment (wuziheng) [inline scripts/easyphoto_tryon_infer.py:286]: here is a bug,should fix
